core/src/main/java/org/infinispan/xsite/events/XSiteEventSender.java

```diff
@@ -0,0 +1,64 @@
+package org.infinispan.xsite.events;
+
+import java.lang.invoke.MethodHandles;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.infinispan.remoting.transport.Transport;
+import org.infinispan.util.ByteString;
+import org.infinispan.util.logging.Log;
+import org.infinispan.util.logging.LogFactory;
+import org.infinispan.xsite.XSiteBackup;
+import org.infinispan.xsite.commands.remote.XSiteRemoteEventCommand;
+
+/**
+ * A collector of events to be sent to the remote site.
+ * <p>
+ * This class implements {@link AutoCloseable} so it can be used with try-with-resources. The {@link #close()} methods sends the events.
+ *
+ * @since 15.0
+ */
+public class XSiteEventSender implements AutoCloseable {
+
+   private static final Log log = LogFactory.getLog(MethodHandles.lookup().lookupClass());
+   private final Transport transport;
+   private final Map<ByteString, EventList> eventsToRemoteSites;
+
+   public XSiteEventSender(Transport transport) {
+      eventsToRemoteSites = new ConcurrentHashMap<>(4);
+      this.transport = transport;
+   }
+
+   public void addEventToSite(ByteString site, XSiteEvent event) {
+      log.debugf("Added event %s to %s", event, site);
+      eventsToRemoteSites.computeIfAbsent(site, EventList::new).add(event);
+   }
+
+   @Override
+   public void close() throws Exception {
+      log.debug("Flushing events");
+      for (var eventList : eventsToRemoteSites.values()) {
+         var cmd = new XSiteRemoteEventCommand(eventList.events);
+         var backup = new XSiteBackup(eventList.site.toString(), false, 10000);
+         transport.backupRemotely(backup, cmd);
```

Comment:
Just curious. As we're sending event cmds individually, is there any benefit to batch events and only sending on close? Are you hoping to take advantage of JGroups batching here, or is the idea to allow some Infinispan level of cmd batching in the future?

Reply:
Each command goes through a different site and the JGroups batching is per destination node. Note that RELAY2 picks a SiteMaster at random for each message and it is impossible to take advantage of JGroups batching, except when `max_site_master=1`.
Hope it explains why I'm sending a single message with a list of events instead of one message per event.

Reply:
That's much clearer, thanks.